File: DevExpress-Demo/VB/SchedulingDemo.Wpf/ViewModels/Highlighted/OnDemandDataLoadingViewModel.vb
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.Native
Imports DevExpress.Xpf.Scheduling
Imports System.Collections.ObjectModel
Imports System.Linq
Imports System.Threading.Tasks
Imports CreateSourceObjectEventArgs = DevExpress.Xpf.Scheduling.CreateSourceObjectEventArgs
#If DXCORE3
using Microsoft.EntityFrameworkCore;
#Else
Imports System.Data.Entity

#End If
Namespace SchedulingDemo.ViewModels

    Public Class OnDemandDataLoadingViewModel

        Const millisecondsDelay As Integer = 2000

        Public Sub New()
            If ViewModelBase.IsInDesignMode Then Return
            DBInitializer.Init()
            Using dbContext = New SchedulingContext()
                Resources = dbContext.ResourceEntities.ToObservableCollection()
            End Using
        End Sub

        Public Overridable Property Resources As ObservableCollection(Of ResourceEntity)

        Public Property Delay As Boolean

        Public Sub ProcessChanges(ByVal args As AppointmentCRUDEventArgs)
            Using dbContext = New SchedulingContext()
                dbContext.AppointmentEntities.AddRange(args.AddToSource.[Select](Function(x) CType(x.SourceObject, AppointmentEntity)))
                Dim updatedAppts = args.UpdateInSource.[Select](Function(x) CType(x.SourceObject, AppointmentEntity))
                Dim updatedApptIds = updatedAppts.[Select](Function(x) x.Id).ToArray()
                Dim apptsToUpdate = dbContext.AppointmentEntities.Where(Function(x) updatedApptIds.Contains(x.Id))
                For Each appt In updatedAppts
                    AppointmentEntityHelper.CopyProperties(appt, apptsToUpdate.First(Function(x) x.Id = appt.Id))
                Next

                Dim deletedApptIds = args.DeleteFromSource.[Select](Function(x) CType(x.SourceObject, AppointmentEntity).Id).ToArray()
                Dim apptsToDelete = dbContext.AppointmentEntities.Where(Function(x) deletedApptIds.Contains(x.Id)).ToArray()
                dbContext.AppointmentEntities.RemoveRange(apptsToDelete)
                dbContext.SaveChanges()
            End Using
        End Sub

        Public Sub CreateSourceObject(ByVal args As CreateSourceObjectEventArgs)
            If args.ItemType = ItemType.AppointmentItem Then args.Instance = New AppointmentEntity()
        End Sub

        Public Sub FetchAppointments(ByVal args As FetchDataEventArgs)
            args.AsyncResult = GetAppointments(args)
        End Sub

        Private Async Function GetAppointments(ByVal args As FetchDataEventArgs) As Task(Of Object())
            If Delay Then Await Task.Delay(millisecondsDelay)
            Using dbContext = New SchedulingContext()
                Return Await dbContext.AppointmentEntities.Where(args.GetFetchExpression(Of AppointmentEntity)()).ToArrayAsync()
            End Using
        End Function
    End Class
End Namespace

File: DevExpress-Demo/CS/SchedulingDemo.Wpf/ViewModels/Highlighted/OnDemandDataLoadingViewModel.cs
using DevExpress.Mvvm;
using DevExpress.Mvvm.Native;
using DevExpress.Xpf.Scheduling;
using System.Collections.ObjectModel;
using System.Linq;
using System.Threading.Tasks;
using CreateSourceObjectEventArgs = DevExpress.Xpf.Scheduling.CreateSourceObjectEventArgs;
#if DXCORE3
using Microsoft.EntityFrameworkCore;
#else
using System.Data.Entity;
#endif

namespace SchedulingDemo.ViewModels {
    public class OnDemandDataLoadingViewModel {
        const int millisecondsDelay = 2000;
        public OnDemandDataLoadingViewModel() {
            if(ViewModelBase.IsInDesignMode)
                return;
            DBInitializer.Init();
            using(var dbContext = new SchedulingContext()) {
                Resources = dbContext.ResourceEntities.ToObservableCollection();
            }
        }
        public virtual ObservableCollection<ResourceEntity> Resources { get; set; }
        public bool Delay { get; set; }

        public void ProcessChanges(AppointmentCRUDEventArgs args) {
            using(var dbContext = new SchedulingContext()) {
                dbContext.AppointmentEntities.AddRange(args.AddToSource.Select(x => (AppointmentEntity)x.SourceObject));

                var updatedAppts = args.UpdateInSource.Select(x => (AppointmentEntity)x.SourceObject);
                var updatedApptIds = updatedAppts.Select(x => x.Id).ToArray();
                var apptsToUpdate = dbContext.AppointmentEntities.Where(x => updatedApptIds.Contains(x.Id));
                foreach(var appt in updatedAppts)
                    AppointmentEntityHelper.CopyProperties(appt, apptsToUpdate.First(x => x.Id == appt.Id));

                var deletedApptIds = args.DeleteFromSource.Select(x => ((AppointmentEntity)x.SourceObject).Id).ToArray();
                var apptsToDelete = dbContext.AppointmentEntities.Where(x => deletedApptIds.Contains(x.Id)).ToArray();
                dbContext.AppointmentEntities.RemoveRange(apptsToDelete);

                dbContext.SaveChanges();
            }
        }
        public void CreateSourceObject(CreateSourceObjectEventArgs args) {
            if(args.ItemType == ItemType.AppointmentItem)
                args.Instance = new AppointmentEntity();
        }
        public void FetchAppointments(FetchDataEventArgs args) {
            args.AsyncResult = GetAppointments(args);
        }
        async Task<object[]> GetAppointments(FetchDataEventArgs args) {
            if(Delay)
                await Task.Delay(millisecondsDelay);
            using(var dbContext = new SchedulingContext()) 
                return await dbContext
                    .AppointmentEntities
                    .Where(args.GetFetchExpression<AppointmentEntity>())
                    .ToArrayAsync();
        }
    }
}
